### DISCUSSION TITLE
Is Mapstruct capable of passing a source object to an @AfterMapping method?

### DISCUSSION THREAD

**Discussion by NadChel** [mapstruct/mapstruct#3299] [Q&A]
Consider this code
```java
@Mapper
@RequiredArgsConstructor
public abstract class QuestionCommentMapper {
    protected final QuestionService questionService;

    public abstract QuestionComment dtoAndAuthenticationToQuestionComment(QuestionCommentRequestDto dto,
                                                                          @Context Authentication auth);

    @AfterMapping
    protected void enrichWithOwner(@MappingTarget QuestionComment questionComment, @Context Authentication auth) {
        Account owner = AuthenticationProcessor.extractAccount(auth);
        questionComment.setOwner(owner);
    }

    @AfterMapping
    protected void enrichWithQuestion(@MappingTarget QuestionComment questionComment,
                                      @Context QuestionCommentRequestDto dto) {
        Long questionId = dto.questionId();
        Question question = questionService.getById(questionId);
        questionComment.setQuestion(question);
    }
}
```
Would Mapstruct pass a `QuestionCommentRequestDto` object to the `enrichWithQuestion()` method which is a source in the original mapping method? If not, how can I perform that second "enriching" without forgoing Mapstruct's code generation? If I write anything in the mapping method (e.g. use the `QuestionCommentRequestDto` instance to set `QuestionComment`'s `Question` field, as in my `enrichWithQuestion()` method), Mapstruct won't generate anything, and I'll have to basically write everything manually

**Comment by NadChel** [comment]
It seems it *is*. Here's the generated code (it didn't occur to me to check it at the point of writing this question)
```java
@Override
    public QuestionComment dtoAndAuthenticationToQuestionComment(QuestionCommentRequestDto dto, Authentication auth) {
        if ( dto == null ) {
            return null;
        }

        QuestionComment questionComment = new QuestionComment();

        questionComment.setText( dto.text() );

        enrichWithOwner( questionComment, auth );
        enrichWithQuestion( questionComment, dto ); // ← look at it

        return questionComment;
    }
```
What made me post it was largely the fact that a chat bot said Mapstruct wouldn't pass the source object because

> Any source parameters used in the regular mapping method are not passed to the @AfterMapping method, as they are not needed for any further mapping operations at that point.

**Reply by filiphr** [reply]
As people like to say nowadays chat bots can "hallucinate" i.e. come up with facts that are not true.

Best place to check things like this is our docs. This is explained in [Mapping customization with before-mapping and after-mapping methods](https://mapstruct.org/documentation/stable/reference/html/#customizing-mappings-with-before-and-after).

>Supported parameters:
>
> * A parameter annotated with `@MappingTarget` is populated with the target instance of the mapping.
> * A parameter annotated with `@TargetType` is populated with the target type of the mapping.
> * Parameters annotated with `@Context` are populated with the context parameters of the mapping method.
> * Any other parameter is populated with a source parameter of the mapping
